openvino_xai/methods/black_box/aise.py

```diff
@@ -0,0 +1,221 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import collections
+from typing import Callable, List
+
+import math
+from scipy.optimize import direct, Bounds
+import numpy as np
+import openvino.runtime as ov
+from openvino.runtime.utils.data_helpers.wrappers import OVDict
+
+from openvino_xai.common.utils import IdentityPreprocessFN, infer_size_from_image, scaling, sigmoid
+from openvino_xai.methods.black_box.base import BlackBoxXAIMethod, Preset
+
+
+class GaussianPerturbationMask:
+    def __init__(self, input_size):
+        h = np.linspace(0, 1, input_size[0])
+        w = np.linspace(0, 1, input_size[1])
+        self.h, self.w = np.meshgrid(w, h)
+
+    def _gaussian_2d(self, gauss_params):
+        mh, mw, sigma = gauss_params
+        A = 1 / (2 * math.pi * sigma * sigma)
+        B = (self.h - mh) ** 2 / (2 * sigma ** 2)
+        C = (self.w - mw) ** 2 / (2 * sigma ** 2)
+        return A * np.exp(-(B + C))
+
+    def generate_kernel_mask(self, gauss_param, scale=1.0):
+        kernel = self._gaussian_2d(gauss_param)
+        return (kernel / kernel.max()) * scale
+
+
+class AISE(BlackBoxXAIMethod):
+    """AISE explains classification models in black-box mode using 
+    AISE: Adaptive Input Sampling for Explanation of Black-box Models.
+
+    :param model: OpenVINO model.
+    :type model: ov.Model
+    :param postprocess_fn: Preprocessing function that extract scores from IR model output.
+    :type postprocess_fn: Callable[[OVDict], np.ndarray]
+    :param preprocess_fn: Preprocessing function, identity function by default
+        (assume input images are already preprocessed by user).
+    :type preprocess_fn: Callable[[np.ndarray], np.ndarray]
+    :param device_name: Device type name.
+    :type device_name: str
+    :param prepare_model: Loading (compiling) the model prior to inference.
+    :type prepare_model: bool
+    """
+
+    def __init__(
+        self,
+        model: ov.Model,
+        postprocess_fn: Callable[[OVDict], np.ndarray],
+        preprocess_fn: Callable[[np.ndarray], np.ndarray] = IdentityPreprocessFN(),
+        device_name: str = "CPU",
+        prepare_model: bool = True,
+    ):
+        super().__init__(model=model, preprocess_fn=preprocess_fn, device_name=device_name)
+        self.postprocess_fn = postprocess_fn
+
+        # TODO: resete state ever run
+        self.data_preprocessed = None
+        self.target = None
+        self.num_iterations_per_kernel = None
+        self.kernel_width = None
+        self.seed = None
+
+        self._kernel_params_hist = collections.defaultdict(list)
+        self._pred_score_hist = collections.defaultdict(list)
+        self.input_size = None
+        self._mask_generator = None
+        
+        if prepare_model:
+            self.prepare_model()
```

Comment:
Just an idea, but what do you think about omitting `prepare_model` parameter and prepare the model automatically if `self._model_compiled` is None?

Reply:
there are some cases, mostly for internal testing convenience, when model should not be prepared